sky/utils/controller_utils.py

```diff
@@ -160,40 +184,80 @@ def from_type(cls, controller_type: str) -> Optional['Controllers']:
+    # characters from the previous message.
+    empty_str = ' ' * 5
+    aws_dependencies_installation = (
+        'pip list | grep boto3 > /dev/null 2>&1 || '
+        'pip install "urllib3<2" awscli>=1.27.10 botocore>=1.29.10 '
+        'boto3>=1.26.1 "colorama<0.4.5" > /dev/null 2>&1')
+    for cloud in sky_check.get_cached_enabled_clouds_or_refresh():
+        if isinstance(
+                clouds,
+            (clouds.Lambda, clouds.SCP, clouds.Fluidstack, clouds.Paperspace)):
+            # no need to install any cloud dependencies for lambda, scp,
+            # fluidstack and paperspace
+            continue
+        if cloud.is_same_cloud(clouds.AWS()):
+            commands.append(f'echo -n "{prefix_str}AWS{empty_str}" && ' +
+                            aws_dependencies_installation)
+        elif cloud.is_same_cloud(clouds.Azure()):
+            commands.append(
+                f'echo -en "\\r{prefix_str}Azure{empty_str}" && '
+                'pip list | grep azure-cli > /dev/null 2>&1 || '
+                'pip install "azure-cli>=2.31.0" azure-core '
+                '"azure-identity>=1.13.0" azure-mgmt-network > /dev/null 2>&1')
+        elif cloud.is_same_cloud(clouds.GCP()):
+            commands.append(
+                f'echo -en "\\r{prefix_str}GCP{empty_str}" && '
+                'pip list | grep google-api-python-client > /dev/null 2>&1 || '
+                'pip install "google-api-python-client>=2.69.0" '
+                '> /dev/null 2>&1')
+            # Have to separate the installation of google-cloud-storage from
+            # above because for a VM launched on GCP, the VM may have
+            # google-api-python-client installed alone.
+            commands.append(
+                'pip list | grep google-cloud-storage > /dev/null 2>&1 || '
+                'pip install google-cloud-storage > /dev/null 2>&1')
+            commands.append(f'{gcp.GOOGLE_SDK_INSTALLATION_COMMAND}')
+        if controller == Controllers.JOB_CONTROLLER:
+            if cloud.is_same_cloud(clouds.IBM()):
+                commands.append(
+                    f'echo -en "\\r{prefix_str}IBM{empty_str}" '
+                    '&& pip list | grep ibm-cloud-sdk-core > /dev/null 2>&1 || '
+                    'pip install ibm-cloud-sdk-core ibm-vpc '
+                    'ibm-platform-services ibm-cos-sdk > /dev/null 2>&1')
+            elif cloud.is_same_cloud(clouds.OCI()):
+                commands.append(f'echo -en "\\r{prefix_str}OCI{empty_str}" && '
+                                'pip list | grep oci > /dev/null 2>&1 || '
+                                'pip install oci > /dev/null 2>&1')
+            elif isinstance(cloud, clouds.Kubernetes):
+                commands.append(
+                    f'echo -en "\\r{prefix_str}Kubernetes{empty_str}" && '
+                    'pip list | grep kubernetes > /dev/null 2>&1 || '
+                    'pip install "kubernetes>=20.0.0" > /dev/null 2>&1')
```

Comment:
should we install k8s for serve controller as well?

Reply:
Seems our serving on kubernetes PR has not merged yet #3377?

Reply:
Oh got it. Then it looks good to me :)